pandas/core/indexes/datetimelike.py

```diff
@@ -273,12 +273,22 @@ def map(self, mapper, na_action=None):
         except Exception:
             return self.astype(object).map(mapper)
 
-    def sort_values(self, return_indexer=False, ascending=True):
+    def sort_values(
+        self, return_indexer=False, ascending=True, key: Optional[Callable] = None
+    ):
         """
         Return sorted copy of Index.
         """
+        if not isinstance(self, Index):
+            raise TypeError("sort_values must be called on an Index object")
+
+        if key:
```

Comment:
I would wrap these up in an ``ensure_key_mapped(self, ey)`` and put this function in pandas/core/sorting.py as this is used in multiple places.

Reply:
i wouldn't use map either rather a think a list-comprehension is fine (you need to mask the null values)

Reply:
I wanted to have an `ensure_key_mapped` function, but it has to work on a lot of different kinds of objects (indices, frames, series, numpy arrays, etc.) How would you handle this? Right now I have a function

```python
def ensure_key_mapped(values, key):
    from pandas.core.generic import NDFrame
    from pandas.core.index import Index

    if not key:
        return values
        
    elif isinstance(values, NDFrame) or isinstance(values, Index):
        return values.map(key, na_action='ignore')
    elif isinstance(values, list):
        return [key(value) for value in values]
    elif isinstance(values, np.ndarray):
        def map_f(values, f):
            return lib.map_infer_mask(values, f, isna(values).view(np.uint8))

        return map_f(values, key)
    else:
        raise TypeError(f"Could not map key to object of type {type(values)}")
```

that works, but it's a little janky.

Reply:
list and ndarray branches can be combined 
you don’t need to use .map at all
so i think you can combined almost everything 

put this in place then keep trying to simplify

Reply:
@jreback is there a reason to avoid `.map`? It seems like it simplifies the implementation, since there's a lot of logic in `Index::map` and `IndexOpsMixin:_map_values` that I don't want to have to reproduce here. In fact, the logic at `core/base.py:1168` seems to be exactly what I want, and that gets called by `Series::map` and `Index::map`.

Reply:
@jreback Everything else is done – I can try and change it to avoid map, but that seems like it means coding separate logic for all the different `Index` subclasses.

Reply:
ok, if you can consolidate the code then will have a look

Reply:
@jreback take a look at the current consolidated version with `core/sorting.py` and `ensure_key_mapped`.